CDCgov/prime-reportstream
# Universal Pipeline filter logging improvements

This PR makes adjustments to the way that filter results are logged in the Universal Pipeline. There were previously several cases where the logged filter message would not have sufficient information to identify why a result was filtered out, and those should be a bit more clear with these changes. For details on the updated behavior, see the Markdown file in the changeset.

Test Steps:
There are several different cases that can/should be tested here, I'll list a few scenarios.

Default filter logging:
1. Delete a receiver's quality filter and reload the credentials.
2. Send a report that matches the receiver's jurisdictional filter, but which fails the default quality filter. (Removing the new lines from [this file](https://github.com/CDCgov/trusted-intermediary/blob/main/examples/fhir/lab_order.json) is an example of a report which does not pass the quality filter)
3. Query the submission history for that report id.
4. Verify that the filter logs in the response include an indication that the default filter was used, and that only the predicate(s) from the default quality filter that evaluate to false are listed.

Reversed filter logging

1. Delete a receiver's quality filter, set `reverseTheQualityFilter` to true, and reload the credentials.
2. Send a report that matches the receiver's jurisdictional filter, and which passes the default quality filter (meaning it will fail a reversed default quality filter).
3. Query the submission history for that report id.
4. Verify that the filter logs in the response include an indication that the default filter was used and the filter was reversed, and all predicates from the default quality filter are listed.

Exception filter logging

1. Set a receiver's quality filter to include some predicate(s) which will cause a schema exception.
2. Send a report that matches the receiver's jurisdictional filter, and which will exercise the the exception-causing filter.
3. Query the submission history for that report id.
4. Verify that the filter logs in the response include an indication that the filter led to an exception, and all predicates which lead to exceptions are listed.

## Changes
- Edit `evaluateFilterCondition` to return a Boolean and a String so that calling functions have access to a String description of the failing filters (which allows for the printed list of predicates to be different than the entire list for the filter, also allows for other indicators for reversed filters or filters causing exceptions)
- Update tests to reflect `evaluateFilterCondition` return type changes, and to remove kotlin warnings with minor changes
- Remove the application of default filters "by lazy" in a simple Boolean form, and instead allow returning of default filters in `getQualityFilters` and `getProcessingModeFilter` (simplifies `evaluateFilterCondition` having access to predicates within the default filter)
- Adjust logging in `evaluateFilterAndLogResult` to identify whether or not the filter applies was a default filter and add to the description of the failing filters to indicate as such

## Checklist

### Testing

## Linked Issues
- Fixes #8915
